### ISSUE TITLE
False Positive for ST11 when joining tables from CTE

### ISSUE THREAD

**Issue by g-campillo** [sqlfluff/sqlfluff#6636]
### Search before asking

- [x] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

Running sqlfluff lint on a specific model raising false positive warning for ST11:

```text
ST11 | Joined table table name> not referenced elsewhere in query [structure.unused_join]
```

Please note that this is slightly different than https://github.com/sqlfluff/sqlfluff/issues/6511 but different as we don't use any of the columns from the table we are joining in our select statement and yet we get the same error

### Expected Behaviour

sqlfluff should not raise this error for the given table(s) as the table reference is then used to join with another table in that model

### Observed Behaviour

```text
ST11 | Joined table table name> not referenced elsewhere in query [structure.unused_join]
```

### How to reproduce

Run `sqlfluff lint` on the following DBT model:

```sql
with customer_info as (
  select
      *
  from {{ ref('customer_info') }}
),

order_details as (
  select
      customer_id,
      created_by,
      created_date
  from {{ ref('order_details') }} as record_details
)

select
  customer.username
from customer_info as customer
left join order_details as orders
  on customer.id = orders.customer_id


```

Again, note that no columns from the order_details CTE were used in the final select statement. The order_details CTE was only used for a join

### Dialect

Redshift

### Version

`sqlfluff, version 3.3.1`
`sqlfluff-templater-dbt, version 3.3.1`
`Python 3.12.7`

DBT Packages:
```text
dbt--labs/codegen -> 0.12.1
dbt-labs/dbt_utls -> 1.3.0
dbt-labs/audit_helper -> 0.12.0
dbt-labs/redshift -> 0.9.0
dbt-labs/dbt_external_tables -> 0.9.0
calogica/dbt_expectations -> 0.10.4
```

### Configuration

.sqlfluff
```text
[sqlfluff]
templater = dbt
dialect = redshift
runaway_limit = 10
indent_unit = space
exlucde_rules = LT07,L029,L051,L059
max_line_length = 120
large_file_skip_byte_limit = 0

[sqlfluff:rules]
allow_scalar = False
single_table_references = unqualified

[sqlfluff:indentation]
tab_space_size = 4
indent_unit = space

[capitalisation.keywords]
capitalisation_policy = lower

[sqlfluff:templater:jinja]
apply_dbt_builtins = True

[sqlfluff:templater:dbt]
project_dir = src
profile = data
target = default

[sqlfluff:layout:type:comma]
spacing_before = touch
line_position = trailing

[sqlfluff:rules:capitalisation.keywords]
capitalisation_policy = lower

[sqlfluff:rules:aliasing.table]
aliasing = explicit

[sqlfluff:rules:aliasing.column]
aliasing = explicit

[sqlfluff:rules:aliasing.expression]
allow_scalar = False

[sqlfluff:rules:capitalisation.identifiers]
extended_capitalisation_policy = lower

[sqlfluff:rules:capitalisation.functions]
capitalisation_policy = lower
extended_capitalisation_policy = lower

[sqlfluff:rules:capitalisation.literals]
capitalisation_policy = lower

[sqlfluff:rules:ambiguous.column_references]
group_by_and_order_by_style = explicit
```

.sqlfluffignore:
```text
**/target/
**/dbt_packages/
**/macros/
```

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [x] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by keraion** [comment]
I'm trying to understand where you say this:
> sqlfluff should not raise this error for the given table(s) as the table reference is then used to join with another table in that model

The rule is currently flagging where your SQL query has a left join, but the right-side table is never referenced. While this may impact the cardinality of the result set, the rule appears to be flagging this appropriately. If any of the columns in the right-side table were used at any point in the query it would not flag this.

If you need the cardinality difference, I'd suggest adding the `-- noqa: ST11` as noted in the rule description.

If I've missed anything, please let me know.
